vhost-user-backend/src/handler.rs

```diff
@@ -6,10 +6,14 @@
 use std::error;
 use std::fs::File;
 use std::io;
+#[cfg(feautre = "postcopy")]
```

Comment:
Shouldn't this be `feature`? :thinking: Shouldn't this just be a no-op as it is?

Reply:
Great catch!

I think we didn't realize it, because we don't have any tests where we call these new APIs.

Maybe we should add a negative test (expecting an error if our `MockVhostBackend` doesn't implement them) in `vhost-user-backend/tests/vhost-user-server.rs`. That way we can at least test the API and the feature.

Reply:
Thx for catching this. Fixed now. With the type the `cfg` just disables whatever is under it (so behaves same as if feature is not enabled).
I did manual testing with Firecracker where I replaced this crate with my fork and triggered all APIs from there for qemu backend.
`MockVhostBackend` in the crate will not be able to test new APIs, because it is the handler that implement them. I don's see `MockVhostBackend` creating handler and testing anything with it.

Reply:
> `MockVhostBackend` in the crate will not be able to test new APIs, because it is the handler that implement them. I don's see `MockVhostBackend` creating handler and testing anything with it.

Why you need to create a handler, the vhost-user-backend is handling it, the specific device doesn't need to do anything, no?
Isn't this similar to `add_mem_region()` for example?

We call it in the test, but `MockVhostBackend` doesn't implement any callback, since it is handled by `BackendReqHandler::handle_request`.

So, I mean something like this (Note: needs to play with features, etc.):

```diff
diff --git a/vhost-user-backend/tests/vhost-user-server.rs b/vhost-user-backend/tests/vhost-user-server.rs
index d8be86d..caa31c1 100644
--- a/vhost-user-backend/tests/vhost-user-server.rs
+++ b/vhost-user-backend/tests/vhost-user-server.rs
@@ -292,3 +292,14 @@ fn vhost_user_get_inflight(path: &Path, barrier: Arc<Barrier>) {
 fn test_vhost_user_get_inflight() {
     vhost_user_server(vhost_user_get_inflight);
 }
+
+fn vhost_user_postcopy(path: &Path, barrier: Arc<Barrier>) {
+    let mut frontend = setup_frontend(path, barrier);
+
+    frontend.postcopy_end().unwrap();
+}
+
+#[test]
+fn test_vhost_user_postcopy() {
+    vhost_user_server(vhost_user_postcopy);
+}
```

It is working, while `postcopy_advice` return an error, I didn't have time to investigate, maybe I need some special permission to create the `userfaultfd`.

Reply:
Oh ok. Sorry, somehow I totally skipped this layered cake of abstractions used in the tests. Would have never thought that in the `vhost-user-backend/tests` there will be tests for both crates.
Anyway I added tests as you suggested. In order to use `uffd` process need access to `dev/userfaultfd`. Considering tests do pass, it seems they run as root in the container, so everything is fine.

Reply:
> Oh ok. Sorry, somehow I totally skipped this layered cake of abstractions used in the tests. Would have never thought that in the `vhost-user-backend/tests` there will be tests for both crates.

The two crates are very connected to each other (so we moved them to the same workspace), especially to test the backend, we need the frontend to send messages. Since in the end our frontend is just a message wrapper, we use it to test the backend.

Reply:
Now it is starting to make some sense, but the amount of indirections and abstractions makes it hard to comprehend what is going on here.

Reply:
yeah, I know :-(